Add FetchX509Bundles rpc (#136)
spiffe/spiffe: standards/SPIFFE_Workload_API.md

    // well as related information like trust bundles and CRLs. As
    // this information changes, subsequent messages will be sent.
    rpc FetchX509SVID(X509SVIDRequest) returns (stream X509SVIDResponse);

    // Fetch trust bundles and CRLs.  Useful for clients that only
    // need to validate SVIDs without obtaining an SVID for themself.

>     // As this information changes, subsequent messages will be sent.

Reviewer: Since this isn't a streaming RPC, I'm not sure what is meant by "subsequent messages will be sent". Would we want to restate this sentence as something like: The return value of this RPC is not necessarily static, clients may periodically use this RPC to retrieve up-to-date contents.

Author: it is a response-streaming RPC...

Reviewer: whoops. n/m!

### 5.2 Federated Bundles

The `FetchX509SVID` RPC will always provide a Trust Bundle for the Trust Domain in which an SVID resides, however, it may also provide bundles for foreign Trust Domains.

> The `FetchX509Bundles` RPC returns a set of Trust Bundles keyed by the SPIFFE ID of the trust domain.  Since this RPC does not return an SVID, all bundles are encoded in the same way in the response, whether they are for the trust domain in which the Server resides or are foreign.

Reviewer 2: ```suggestion
The `FetchX509Bundles` RPC returns a set of Trust Bundles keyed by the SPIFFE ID of the trust domain.  Since this RPC does not return an SVID, all bundles are encoded in the same way in the response, whether they are for the trust domain in which the server resides or are foreign.
```

```protobuf
// The X509SVIDResponse message carries a set of X.509 SVIDs and their
// associated information. It also carries a set of global CRLs, and a

> // map of federated bundles the workload should trust.

Reviewer: nit: this sentence doesn't need the comma.

    bytes bundle = 4;

}

> // The X509BundlesResponse message carries a set of global CRLs, and a

Reviewer: nit: this sentence doesn't need the comma.

// map of trust bundles the workload should trust.
message X509BundlesResponse {

>     // ASN.1 DER encoded

Reviewer: nit: please end this comment with punctuation.

    repeated bytes crl = 1;

    // CA certificate bundles belonging to Trust Domains that the
    // workload should trust, keyed by the SPIFFE ID of the trust
    // domain. Bundles are ASN.1 DER encoded.
    map<string, bytes> bundles = 2;
}
```

All fields in the `X509SVID` message are mandatory, and MUST contain a non-default value. Clients receiving an `X509SVID` message in which any field has a default value SHOULD report an error and discard the message.

The only mandatory field in the `X509SVIDResponse` message is the `svids` field. If the client is not entitled to an SVID, then the server SHOULD respond with the "PermissionDenied" gRPC status code (see the [Error Codes](SPIFFE_Workload_Endpoint.md#6-error-codes) section in the SPIFFE Workload Endpoint specification for more information). The `crl` field and the `federated_bundles` field are optional, and may contain a default value.

> The `X509BundlesResponse` message MUST contain at least one trust bundle.  If the Client is not entitled to recieve any X509 bundles, the server SHOULD respond with the "PermissionDenied" gRPC status code.

Reviewer 2: ```suggestion
The `X509BundlesResponse` message MUST contain at least one trust bundle.  If the client is not entitled to receive any X.509 bundles, the server SHOULD respond with the "PermissionDenied" gRPC status code.
```

### 5.5 Default Values and Redacted Information

SPIFFE Workload API clients may at times encounter fields in the response message that have a default value, or may notice that information included in a previous response is not included in the latest response. For instance, a client may encounter a default value in the `federated_bundles` field after having previously received a federated bundle.

Since every message MUST include the full set of information (see the [Workload API Client and Server Behavior](#51-workload-api-client-and-server-behavior) section), clients SHOULD interpret the absence of data as a redaction. As an example, if a client has loaded a bundle for `spiffe://foo.bar`, and receives a message that does not include a bundle for `spiffe://foo.bar`, then the bundle SHOULD be unloaded.

If the server redacts all SVIDs from a workload, it SHOULD send the "PermissionDenied" gRPC status code (terminating the gRPC response stream). The client SHOULD cease using the redacted SVIDS. The client MAY attempt to reconnect with another call to the `FetchX509SVID` RPC after a backoff.

> If the server redacts all trust bundles from a Client using the `FetchX509Bundles` RPC, it SHOULD send the "PermissionDenied" gRPC status code (terminating the gRPC response stream). The client SHOULD cease using the redacted trust bundles. The client MAY attempt to reconnect with another call to the `FetchX509Bundles` RPC after a backoff.

Reviewer 2: ```suggestion
If the server redacts all trust bundles from a client using the `FetchX509Bundles` RPC, it SHOULD send the "PermissionDenied" gRPC status code (terminating the gRPC response stream). The client SHOULD cease using the redacted trust bundles. The client MAY attempt to reconnect with another call to the `FetchX509Bundles` RPC after a backoff.
```
